exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,490 @@
+    /// - A transaction execution status (see `ExecutionResult` and `TransactionResult` for the
+    ///   details) is stored in the blockchain and can be accessed through api.
+    /// - Blockchain state hash is affected by the transactions execution status.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[macro_use] extern crate exonum;
+    /// #
+    /// use exonum::blockchain::{Transaction, ExecutionResult};
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #
+    /// #       public_key: &PublicKey,
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn execute(&self, fork: &mut Fork) -> ExecutionResult {
+    ///         // Read and/or write into storage.
+    ///         // ...
+    ///
+    ///         // Return execution status.
+    ///         Ok(())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn verify(&self) -> bool { true }
+    /// }
+    /// # fn main() {}
+    fn execute(&self, fork: &mut Fork) -> ExecutionResult;
+}
+
+/// Result of unsuccessful transaction execution.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ExecutionError {
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    code: u8,
+    /// Optional error description.
+    description: Option<String>,
+}
+
+impl ExecutionError {
+    /// Constructs a new `ExecutionError` instance with the given error code.
+    pub fn new(code: u8) -> Self {
+        Self {
+            code,
+            description: None,
+        }
+    }
+
+    /// Constructs a new `ExecutionError` instance with the given error code and description.
+    pub fn with_description(code: u8, description: String) -> Self {
+        Self {
+            code,
+            description: Some(description),
+        }
+    }
+}
+
+/// Type of the transaction error.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TransactionErrorType {
+    /// Panic occurred during transaction execution.
+    Panic,
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    Code(u8),
+}
+
+/// Extended by the framework result of unsuccessful transaction execution.
+///
+/// # Notes:
+///
+/// - Content of `description`' field is excluded from hash calculation (see `StorageValue`
+///   implementation for the details).
+/// - `TransactionErrorType::Panic` is set by the framework if panic is raised during transaction
+///   execution.
+/// - `TransactionError` implements `Display` which can be used for obtaining a simple error
+///   description.
+///
+/// # Examples
+///
+/// ```
+/// # use exonum::blockchain::{ExecutionError, TransactionError};
```

Comment:
You sure this example is appropriate? Unless I'm missing something, it would be rendered as a single line of code telling very little how `TransactionError` can be used.

Reply:
I'm not sure, do you have better ideas?

Reply:
Well, the most complete example I can think of is demonstrating how an `ExecutionError` returned by the service can then be read from the core schema; something like

```rust
message! {
    struct ErrTransaction { }
}

impl Transaction for ErrTransaction {
    // ...
    fn execute(&self, _: &mut Fork) -> ExecutionResult {
         Err(ExecutionError::with_description(1, "Gosh!"))
    }
}

let mut blockchain = ...;
let patch = blockchain.create_patch(...);
blockchain.merge(patch);

let schema = Schema::new(blockchain.snapshot());
let tx_result = schema.transaction_results().get(tx.hash()).unwrap();
let stored_err = tx_result.unwrap_err();
assert_eq!(stored_err.error_type(), TransactionErrorType::Code(1));
assert_eq!(stored_err.description(), Some("Gosh"));
```

Obviously, to make this example actually compile, some hidden code is needed.